src/classes/ATTD_CourseConnectionContact_TDTM.cls

```diff
@@ -0,0 +1,90 @@
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Attendance Events
+* @group-content ../../ApexDocContent/AttendanceEvents.htm
+* @description Makes sure that linked course connection's contact matches contact lookup field
+*/
+public with sharing class ATTD_CourseConnectionContact_TDTM extends TDTM_Runnable {
+    /*******************************************************************************************************
+    * @description Makes sure that linked course connection's contact matches contact lookup field
+    * @param listNew the list of Attendance Events from trigger new. 
+    * @param listOld the list of Attendance Events from trigger old. 
+    * @param triggerAction which trigger event (BeforeInsert, AfterInsert, etc.). 
+    * @param objResult the describe for Attendance Events 
+    * @return dmlWrapper.  
+    ********************************************************************************************************/
+    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
+    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+
+        DmlWrapper dmlWrapper = new DmlWrapper();
+
+        boolean recursionFlag = TDTM_ProcessControl.getRecursionFlag(TDTM_ProcessControl.registeredTrigger.ATTD_CourseConnectionContact_TDTM);
+
+        Map<ID, SObject> newmap = new Map<ID, SObject>(newlist);
+        
+        if (triggerAction == TDTM_Runnable.Action.AfterInsert
+                || triggerAction == TDTM_Runnable.Action.AfterUpdate) {
+                    
+            if (!recursionFlag) {
+                for (Attendance_Event__c attd : [SELECT Id,
+                                                        Contact__c,
+                                                        Course_Connection__c,
+                                                        Course_Connection__r.Contact__c
+                                                FROM Attendance_Event__c 
+                                                WHERE Id IN :newlist]) {
+                    if (attd.Course_Connection__c != null
+                        && attd.Course_Connection__r.Contact__c != null
+                        && attd.Course_Connection__r.Contact__c != attd.Contact__c) {
```

Comment:
@xinnanlisf As per our conversation we could move attd.Course_Connection__r.Contact__c != attd.Contact__c to line 68.